v-next/hardhat-errors/src/errors.ts

```diff
@@ -2,45 +2,92 @@ import { CustomError } from "@nomicfoundation/hardhat-utils/error";
 
 import { ERRORS, ErrorDescriptor } from "./descriptors.js";
 
-export const ERROR_PREFIX = "HHE";
-
-const IS_HARDHAT_ERROR_PROPERTY_NAME = "_isHardhatError";
-
 export type ErrorMessageTemplateValue =
   | string
   | number
   | boolean
   | bigint
   | undefined
   | null
+  | ErrorMessageTemplateValue[]
   | { toString(): string };
 
-export type ErrorMessageTemplateArguments = Record<
-  string,
-  ErrorMessageTemplateValue
->;
+export type MessagetTemplateArguments<MessageTemplateT extends string> =
+  MessageTemplateT extends `${string}{${infer Tag}}${infer Rest}`
+    ? {
+        [K in
+          | Tag
+          | keyof MessagetTemplateArguments<Rest>]: ErrorMessageTemplateValue;
+      }
+    : {};
+
+export type HardhatErrorConstructorArguments<
+  ErrorDescriptorT extends ErrorDescriptor,
+> = keyof MessagetTemplateArguments<
+  ErrorDescriptorT["messageTemplate"]
+> extends never
+  ? [ErrorDescriptorT, Error?]
+  : [
+      ErrorDescriptorT,
+      MessagetTemplateArguments<ErrorDescriptorT["messageTemplate"]>,
+      Error?,
+    ];
 
-export class HardhatError extends CustomError {
+export const ERROR_PREFIX = "HHE";
+
+const IS_HARDHAT_ERROR_PROPERTY_NAME = "_isHardhatError";
+
+export class HardhatError<
+  ErrorDescriptorT extends ErrorDescriptor,
+> extends CustomError {
   public static readonly ERRORS = ERRORS;
 
-  readonly #descriptor: ErrorDescriptor;
+  readonly #descriptor: ErrorDescriptorT;
+
+  readonly #arguments: MessagetTemplateArguments<
+    ErrorDescriptorT["messageTemplate"]
+  >;
 
   constructor(
-    errorDescriptor: ErrorDescriptor,
-    messageArguments: ErrorMessageTemplateArguments = {},
-    parentError?: Error,
+    ...[
+      errorDescriptor,
+      messageArgumentsOrParentError,
+      parentError,
+    ]: HardhatErrorConstructorArguments<ErrorDescriptorT>
   ) {
     const prefix = `${getErrorCode(errorDescriptor)}: `;
 
-    const formattedMessage = applyErrorMessageTemplate(
-      errorDescriptor.messageTemplate,
-      messageArguments,
+    const formattedMessage =
+      messageArgumentsOrParentError === undefined ||
+      messageArgumentsOrParentError instanceof Error
+        ? errorDescriptor.messageTemplate
+        : applyErrorMessageTemplate(
+            errorDescriptor.messageTemplate,
+            messageArgumentsOrParentError,
+          );
+
+    super(
+      prefix + formattedMessage,
+      parentError instanceof Error
+        ? parentError
+        : messageArgumentsOrParentError instanceof Error
+          ? messageArgumentsOrParentError
+          : undefined,
     );
 
-    super(prefix + formattedMessage, parentError);
-
     this.#descriptor = errorDescriptor;
 
+    if (
+      messageArgumentsOrParentError === undefined ||
+      messageArgumentsOrParentError instanceof Error
+    ) {
+      this.#arguments = {} as MessagetTemplateArguments<
```

Comment:
Was this the problematic line you where talking about? 😄

Reply:
This one was also problematic, but the one I was confused about was that if you had no variables in your template this would work `new HardhatError(descriptor, {extraField: 123}, cause)`.

The solution was to change the constructor so that now, if you don't have any variable, it is called like this `new HardhatError(descriptor, cause)`